PP-745: Fix for single node job, string type resource encoding.

dilip-krishnan opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-745](https://pbspro.atlassian.net/browse/PP-745)**

#### Problem description
* Single node jobs don't quote JSON formatted strings.

#### Cause / Analysis
* While accumulating resources for a multinode job string type  resources are merge if
they are in json format  and are set within single quotes. But in case of single node job such
accumulation process was not required, due to which string resources having json string were
not quoted. This causes an inconsistent of resource usage log in accounting log. 

#### Solution description
* In case of single node job, in encode_resc_used check for json string resources and set them within single quotes.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

Klovely commented on test/tests/functional/pbs_resc_used_single_node.py:
```diff
@@ -0,0 +1,195 @@
+from tests.functional import *
+
+
+class TestPbsPP745(TestFunctional):
```
We can give a more meaningful class name instead of TestPbs745.

prakashcv13 commented:
@dilip-krishnan, We would also need a external interface design document for these changes.

dilip-krishnan commented:
Hi @prakashcv13  & @Klovely ,
     I have addressed your comments. I am working on the EDD and running the pbs_accumulate.. test case.
Please review the latest changes.

Warm regards
Dilip

Klovely commented:
Changes looks good to me!

dilip-krishnan commented:
Hi @PBSPro/pbspro-maintainers,
         This PR is ready for merge please review the changes.

bhroam commented on test/tests/functional/pbs_resc_used_single_node.py:
```diff
@@ -0,0 +1,234 @@
+            try:
+                self.server.manager(MGR_CMD_DELETE, RSC, id=r, runas=ROOT_USER)
+            except:
+                pass
```
You don't need to remove resources in tearDown().  revert_to_defaults() does that.

dilip-krishnan replied:
I noticed that revert_to_defaults was not able to remove resource created by ptl on my centos 7 dev machine. I had to manually delete and need to restart mom , server services. To restore to defaults.This behaviour is observed on both master and PP-745.

Hence I added additional step, just to make sure to delete resources get removed before and after the test case ran

bhroam commented on test/tests/functional/pbs_resc_used_single_node.py:
```diff
@@ -0,0 +1,234 @@
+            try:
+                self.server.manager(MGR_CMD_DELETE, RSC, id=r, runas=ROOT_USER)
+            except:
+                pass
```
These resources won't exist.  revert_to_defaults() is called in TestFunctional.setUp().  It removes all custom resources.

dilip-krishnan replied:
Same reason as above.

bhroam commented on test/tests/functional/pbs_resc_used_single_node.py:
```diff
@@ -0,0 +1,234 @@
+            except:
+                pass
+
+        self.server.restart()
```
Why restart the server?

dilip-krishnan replied:
Same reason as above.

bhroam commented on test/tests/functional/pbs_resc_used_single_node.py:
```diff
@@ -0,0 +1,234 @@
+
+        self.momA = self.moms.values()[0]
+        self.momA.restart()
+        self.momA.delete_vnode_defs()
```
Why restart the mom before deleting the vnode def file?

dilip-krishnan replied:
Same as above.

bhroam commented on test/tests/functional/pbs_resc_used_single_node.py:
```diff
@@ -0,0 +1,234 @@
+        # Next set some custom resources via qmgr -c 'create resource'
+        attr = {}
+        attr['type'] = 'string'
+        attr['flag'] = 'h'
```
Suggestion: Combine these lines into the initialization of attr.  This means attr = {'type': 'string', 'flag': 'h'}

dilip-krishnan replied:
Done.

bhroam commented on test/tests/functional/pbs_resc_used_single_node.py:
```diff
@@ -0,0 +1,234 @@
+        self.assertEqual(rc, 0)
+
+        attr['type'] = 'string'
+        attr['flag'] = 'h'
```
No need to reset attr.  It's still set from above.

dilip-krishnan replied:
Done

bhroam commented on test/tests/functional/pbs_resc_used_single_node.py:
```diff
@@ -0,0 +1,234 @@
+        self.assertEqual(rc, 0)
+
+        attr['type'] = 'string_array'
+        attr['flag'] = 'h'
```
The flag hasn't changed from above.  No need to reset it.

dilip-krishnan replied:
Done

bhroam commented on test/tests/functional/pbs_resc_used_single_node.py:
```diff
@@ -0,0 +1,234 @@
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+import time
```
Question: Why break this test out from the test plan that tests single and multi node accumulation?  It just proliferates the number of test plans.

dilip-krishnan replied:
Hi @bhroam,
     On my dev Centos 7 machine, i noticed that resources created were not getting cleaned in ptl by revert_to_defaults. Due to which the test case failed if ran multiple time  due resource  Duplicate resource entry already exist error. Also Multi node test case does a minimum 3 mom requirement test in setup and many other multi-node related setup. Hence too much refactoring was required in the existing test to just make single node test case work. Therefore it put in a separate file.

Regards

bhroam replied:
Strange, I've seen PTL remove resources before.  If it isn't removing the resources, then that's a PTL bug.  Please file it.  As for multiple files, I understand your reasoning.  This isn't the first time tests have been split between single/multi-node tests.  I'd rather see them in the same file, but you are correct about the refactoring needed.  I'm fine with it.

bhroam commented:
@dilip-krishnan please squash and rebase for preparation for the final merge.

dilip-krishnan commented:
Hi @bhroam ,
    I have rebased and sqaushed the PR to single commit.

Regards 
Dilip

bhroam commented:
Looks good, merged.
